### PR TITLE
Use argparse for gather command parameters

### PR DESCRIPTION
**Pull request by tadhg-ohiggins**
The Python standard
Library has many things.
Argparse isn't bad.

Switch the gather command over to using argparse for handling its arguments, and add tests.

My main fear here is that there's some obscure invocation I've broken in making this change, so please let me know if I've missed some corner cases.

### CODE REVIEW

File: gather [hunk 1]
```diff
@@ -48,11 +48,11 @@ def run(options=None):
         os.remove(result)
 
     # Opt in to include parent (second-level) domains.
-    include_parents = options.get("include-parents", False)
+    include_parents = options.get("include_parents", False)
 
     # Opt into stripping www. prefixes from hostnames, effectively
     # collapsing www.[host] and [host] into one record.
-    ignore_www = options.get("ignore-www", False)
+    ignore_www = options.get("ignore_www", False)
```
Comment (konklone) [inline gather]: Are hyphens for arguments no good? This will break some invocations, though it's obviously doable to update/notify downstream users if need be.

File: tests/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            'dap': 'https://analytics.usa.gov/data/live/sites-extended.csv',
+            'private': '/path/to/private-research.csv',
+            'parents': 'https://github.com/GSA/data/raw/master/dotgov-domains/current-federal.csv',
+        }
+    ),
+    (
+        "".join([
+            "./gather dap --suffix=.gov --dap=",
+            "https://analytics.usa.gov/data/live/sites-extended.csv",
+            " --ignore-www",
```
Comment (konklone) [inline tests/test_utils.py]: Maybe I'm not intuiting what the change means -- do `--ignore-www` and `-ignore_www` both work now?

File: utils/utils.py [hunk 3]
```diff
@@ -79,6 +82,103 @@ def options():
...
+
+def options_for_gather():
+    # Parse options for the ``gather`` command.
+    set_services = ("censys")
+    services = [s for s in sys.argv[1].split(",") if s not in set_services]
+    parser = build_gather_options_parser(services)
+    parsed, remaining = parser.parse_known_args()
+    for remainder in remaining:
+        if remainder.startswith("--"):
+            raise argparse.ArgumentTypeError(
```
Comment (konklone) [inline utils/utils.py]: I guess I was expecting a little more magic from the options parser, then having to ever check for `.startswith("--")`. Is there something unusual this is trying to do to preserve the `gather` syntax?

File: gather (hunk 1 above)
Comment (tadhg-ohiggins) [reply gather]: The hyphens on the command-line still work fine, it's just that argparse will automatically transform them into underscores so that they work as Python object property names. No change on the user side should be necessary.

File: tests/test_utils.py (hunk 2 above)
Comment (tadhg-ohiggins) [reply tests/test_utils.py]: On the command line only `--ignore-www` will work; `ignore_www` is what the property name will be after argparse processes them.

File: utils/utils.py (hunk 3 above)
Comment (tadhg-ohiggins) [reply utils/utils.py]: This check is probably being overly-cautious, but because we're dynamically creating new options here based on what gatherer is specified (for example, `dap` means that `--dap` is required), we have to account for options that the parser doesn't recognize, and this clause is trying to catch some stray flag that isn't a gatherer that the user might have entered.

It's possible that instead of dynamically turning gatherers into arguments, what we should be doing is scanning the `gatherers` directory for its contents first and constructing arguments out of that. I'm not yet sure about which of those would be better. It wouldn't be that big a change to switch over, though.

Comment (konklone) [reply utils/utils.py]: That makes sense. Thanks for the explanation. I'm not sure which is better either, and this works for now either way.
